### PR TITLE
Selecting non-reputable nodes should not result in excessive selection

### PR DESCRIPTION
**Pull request by kaloyan-raev**
Fixes https://storjlabs.atlassian.net/browse/V3-1487

Thanks for submitting a PR!

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: pkg/overlay/cache.go [hunk 1]
```diff
@@ -225,32 +225,32 @@ func (cache *Cache) FindStorageNodesWithPreferences(ctx context.Context, req Fin
 		reputableNodeCount = req.RequestedCount
 	}
 
-	auditCount := preferences.AuditCount
-	if auditCount < preferences.NewNodeAuditThreshold {
-		auditCount = preferences.NewNodeAuditThreshold
-	}
-
-	reputableNodes, err := cache.db.SelectStorageNodes(ctx, reputableNodeCount, &NodeCriteria{
+	newNodeCount := int(float64(reputableNodeCount) * preferences.NewNodePercentage)
```
Comment (mniewrzal) [inline pkg/overlay/cache.go]: can you check if `preferences.NewNodePercentage > 0` otherwise set `newNodeCount = 0`?

Comment (kaloyan-raev) [reply pkg/overlay/cache.go]: Makes sense. Done.

**Comment by littleskunk** [comment]
As far as I remember this PR didn't fix the issue and we used an workaround for release-alpha8.

The last error was:
```
root@kali:~/storj# head -c 1M </dev/urandom | uplink put s3://littleskunk/1M
Fatal error: segment error: ecclient error: duplicated nodes are not allowed; segment error: ecclient error: duplicated nodes are not allowed
```

File: pkg/overlay/cache.go [hunk 2]
```diff
@@ -245,25 +265,11 @@ func (cache *Cache) FindStorageNodesWithPreferences(ctx context.Context, req Fin
 		return nil, err
```
[the remark this answers is not in the document]
Comment (kaloyan-raev) [reply pkg/overlay/cache.go]: Done with https://github.com/storj/storj/pull/1680/commits/2e56e7393860e1b428896196c0ac0f5229c07ab7

File: pkg/overlay/cache.go [hunk 3]
```diff
@@ -243,35 +270,19 @@ func (cache *Cache) FindStorageNodesWithPreferences(ctx context.Context, req Fin
...
-		FreeDisk:      req.FreeDisk,
-
-		AuditThreshold: preferences.NewNodeAuditThreshold,
-
-		Excluded: req.ExcludedNodes,
+		Excluded: excluded,
 
 		MinimumVersion: preferences.MinimumVersion,
 	})
 	if err != nil {
```
Comment (egonelbre) [inline pkg/overlay/cache.go]: I'm wondering whether we should fall back to selecting any node when there aren't enough nodes at this point?

Comment (kaloyan-raev) [reply pkg/overlay/cache.go]: No. The callers of this method already expect to get the exact number of the requested number of nodes. No less, no more.

Comment (egonelbre) [reply pkg/overlay/cache.go]: I was thinking of when there aren't enough storage nodes that are reputable.

Comment (kaloyan-raev) [reply pkg/overlay/cache.go]: Do you mean to complete the number with new vetted nodes?

I am not sure if this makes sense. We should manage our vetting process in a way that we have at least 90-95 reputable nodes in the network.

It is worth discussing this, but I think is out of scope for this PR.

Comment (egonelbre) [reply pkg/overlay/cache.go]: Sure let's bring the discussion to design-drafts.

**Pull request merged**